### PR TITLE
Create Azure OpenAI detector

### LINKED ISSUE #1939: Add OpenAI Azure Detector
Please review the [Community Note](https://github.com/trufflesecurity/trufflehog/blob/main/.github/community_note.md) before submitting

## Description

The current [OpenAI detector](https://github.com/trufflesecurity/trufflehog/blob/b8724e87e6c7f6785bee707a82172d03e3006efb/pkg/detectors/openai/openai.go#L23) handles API keys for `https://api.openai.com`, however, in my experience API keys for [Microsoft Azure Endpoints](https://github.com/openai/openai-python#microsoft-azure-endpoints) have a different format.

For example:
```python
import openai
openai.api_type = "azure"
openai.api_base = "https://example-endpoint.openai.azure.com"
openai.api_version = "2023-05-15"
openai.api_key = "[a-f0-9]{32}" 
```

### Preferred Solution

Add an additional detector for OpenAI Azure, if the key format is indeed different.

### Additional Context
N/A

#### References

N/A

### PR DESCRIPTION
**Pull request by rgmz**
### Description:
This fixes #1939.

### Checklist:
* [x] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/detectors/azure_openai/azure_openai.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			}
+		}
+
+		// JSON unmarshal doesn't check whether the structure actually matches.
+		if deployments.Object == "" {
+			return false, nil, nil
+		}
+
+		extraData := map[string]string{
+			"base_url": "https://" + baseUrl,
```
Comment (rgmz) [inline pkg/detectors/azure_openai/azure_openai.go:140]: Should the URL be extradata or `rawv2`?

File: pkg/engine/ahocorasick/ahocorasickcore.go [hunk 2]
```diff
@@ -1,9 +1,10 @@
-package engine
+package ahocorasick
```
Comment (rgmz) [inline pkg/engine/ahocorasick/ahocorasickcore.go:1]: https://github.com/trufflesecurity/trufflehog/pull/2342#discussion_r1467906169

File: pkg/detectors/azure_openai/azure_openai.go (hunk 1 above)
Comment (dustin-decker) [reply pkg/detectors/azure_openai/azure_openai.go:140]: We use rawV2 to deduplicate findings, so rawV2 should have the key+url. Extra data is for including any extra helpful information.

Comment (rgmz) [reply pkg/detectors/azure_openai/azure_openai.go]: Gotcha. I've updated this to pass it in RawV2.

**Comment by dustin-decker** [comment]
Thank you. I've submitted a request for using Azure Open AI so we can make test credentials.

**Comment by ahrav** [comment]
> Thank you. I've submitted a request for using Azure Open AI so we can make test credentials.

@dustin-decker do you know if these have been setup?

File: pkg/detectors/azure_openai/azure_openai.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
+package azure_openai
+
+import (
```
Comment (ahrav) [inline pkg/detectors/azure_openai/azure_openai.go:3]: nit: Can we sort these.
```
import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	regexp "github.com/wasilibs/go-re2"

	"github.com/trufflesecurity/trufflehog/v3/pkg/cache/simple"
	"github.com/trufflesecurity/trufflehog/v3/pkg/common"
	logContext "github.com/trufflesecurity/trufflehog/v3/pkg/context"
	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors"
	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/detectorspb"
)
```

File: pkg/detectors/azure_openai/azure_openai.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	// Most endpoints require additional info, e.g., deployment name, which complicates verification.
+	// This may be retired in the future, so we should look for another candidate.
+	// https://learn.microsoft.com/en-us/answers/questions/1371786/get-azure-openai-deployments-in-api
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("https://%s/openai/deployments?api-version=2023-03-15-preview", baseUrl), nil)
+	if err != nil {
+		return false, nil, nil
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("api-key", token)
```
Comment (ahrav) [inline pkg/detectors/azure_openai/azure_openai.go:116]: question: Just double checking this header name is `api-key` and not `Api-Key`?

**Review by ahrav — APPROVED** [review]
Thanks for adding this!

File: pkg/detectors/azure_openai/azure_openai.go (hunk 3 above)
Comment (rgmz) [reply pkg/detectors/azure_openai/azure_openai.go:3]: Done.

**Pull request merged**